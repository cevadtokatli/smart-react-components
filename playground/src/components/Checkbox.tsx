import { Checkbox, CheckboxGroup } from '@smart-react-components/ui'
import React from 'react'

export const Palette = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <CheckboxGroup active={active} setActive={setActive}>
      <Checkbox value="primary" label="Primary" />
      <Checkbox value="secondary" palette="secondary" label="Secondary" />
      <Checkbox value="success" palette="success" label="Success" />
      <Checkbox value="danger" palette="danger" label="Danger" />
      <Checkbox value="warning" palette="warning" label="Warning" />
      <Checkbox value="info" palette="info" label="Info" />
      <Checkbox value="dynamic" palette="dynamic" label="Dynamic" />
      <Checkbox value="!dynamic" palette="!dynamic" label="!Dynamic" />
    </CheckboxGroup>
  )
}

export const Soft = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <CheckboxGroup active={active} setActive={setActive}>
      <Checkbox value="primary" label="Primary" isSoft />
      <Checkbox value="secondary" palette="secondary" label="Secondary" isSoft />
      <Checkbox value="success" palette="success" label="Success" isSoft />
      <Checkbox value="danger" palette="danger" label="Danger" isSoft />
      <Checkbox value="warning" palette="warning" label="Warning" isSoft />
      <Checkbox value="info" palette="info" label="Info" isSoft />
      <Checkbox value="dynamic" palette="dynamic" label="Dynamic" isSoft />
      <Checkbox value="!dynamic" palette="!dynamic" label="!Dynamic" isSoft />
    </CheckboxGroup>
  )
}

export const Outline = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <CheckboxGroup active={active} setActive={setActive}>
      <Checkbox value="primary" label="Primary" isOutline />
      <Checkbox value="secondary" palette="secondary" label="Secondary" isOutline />
      <Checkbox value="success" palette="success" label="Success" isOutline />
      <Checkbox value="danger" palette="danger" label="Danger" isOutline />
      <Checkbox value="warning" palette="warning" label="Warning" isOutline />
      <Checkbox value="info" palette="info" label="Info" isOutline />
      <Checkbox value="dynamic" palette="dynamic" label="Dynamic" isOutline />
      <Checkbox value="!dynamic" palette="!dynamic" label="!Dynamic" isOutline />
    </CheckboxGroup>
  )
}

export const OutlineSoft = () => {
  const [active, setActive] = React.useState(() => ['primary'])

  return (
    <CheckboxGroup active={active} setActive={setActive}>
      <Checkbox value="primary" label="Primary" isOutline isSoft />
      <Checkbox value="secondary" palette="secondary" label="Secondary" isOutline isSoft />
      <Checkbox value="success" palette="success" label="Success" isOutline isSoft />
      <Checkbox value="danger" palette="danger" label="Danger" isOutline isSoft />
      <Checkbox value="warning" palette="warning" label="Warning" isOutline isSoft />
      <Checkbox value="info" palette="info" label="Info" isOutline isSoft />
      <Checkbox value="dynamic" palette="dynamic" label="Dynamic" isOutline isSoft />
      <Checkbox value="!dynamic" palette="!dynamic" label="!Dynamic" isOutline isSoft />
    </CheckboxGroup>
  )
}

export const Size = () => {
  const [active, setActive] = React.useState(() => ['medium'])

  return (
    <CheckboxGroup active={active} setActive={setActive}>
      <Checkbox value="small" size="small" label="Small" />
      <Checkbox value="medium" label="Medium" />
      <Checkbox value="large" size="large" label="Large" />
    </CheckboxGroup>
  )
}

export const Shape = () => {
  const [active, setActive] = React.useState(() => ['rectangle'])

  return (
    <CheckboxGroup active={active} setActive={setActive}>
      <Checkbox value="rectangle" label="Rectangle" />
      <Checkbox value="rounded" shape="rounded" label="Rounded" />
    </CheckboxGroup>
  )
}

export const Block = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <CheckboxGroup active={active} setActive={setActive} isBlock isRequired>
      <Checkbox value="item-1" label="Item 1" />
      <Checkbox value="item-2" label="Item 2" />
      <Checkbox value="item-3" label="Item 3" />
    </CheckboxGroup>
  )
}
