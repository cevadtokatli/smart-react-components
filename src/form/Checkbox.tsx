import React from "react"
import styled from "styled-components"
import ElementProps from "../element/Props"
import {Size$StrictProps} from "../props/size$-strict-props"
import size from "../constants/form/checkbox-size"
import InputElement from "../element/Input"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, JSXChild, Value, SetValue} from "../props"
import DV from "../default-value"
import useMultiple from "./UseMultiple"
import RadioLabel from "./RadioLabel"
import RadioSpan from "./RadioSpan"

const Rectangle = styled.div``

const StickContainer = styled.div``

const ShortStick = styled.div``

const LongStick = styled.div``

interface InputProps extends ElementProps, Size$StrictProps {
    type$: string
    fill$: boolean
    shape: string
    disabled?: boolean
}

const customKeys = {
    size$: (v, t) => `
        & + ${Rectangle} {
            width: ${size.rectangle[v]};
            height: ${size.rectangle[v]};

            > ${StickContainer} {
                left: ${size.stickContainer[v].left};
                top: ${size.stickContainer[v].top};

                > ${ShortStick} {
                    top: ${size.shortStick[v].top};
                }

                > ${LongStick} {
                    top: ${size.longStick[v].top};
                }
            }
        }

        &:checked + ${Rectangle} > ${StickContainer} > {
            ${ShortStick} {
                height: ${size.shortStick[v].height};
            }

            ${LongStick} {
                width: ${size.longStick[v].width};
            }
        }

        ~ ${RadioSpan} {
            font-size: ${t.form.radioSpanFontSize[v]};
        }
    `
}

const Input = styled(InputElement).attrs<InputProps>({customKeys})<InputProps>(({theme,type$,fill$,shape,disabled}) => `
    flex: 0 0 auto;
    display: none;

    + ${Rectangle} {
        position: relative;
        overflow: hidden;
        flex: 0 0 auto;
        box-sizing: border-box;
        border-radius: ${shape != "circle" ? "2px" : "100%"};
        border: solid 2px ${theme.src.type[type$].form.radioBorder};
        background: ${fill$ ? theme.src.type[type$].form.radioBorder : theme.src.type[type$].form.radioBackground};
        transition: 100ms 0s ease-in-out;
        transition-property: border, background;

        > ${StickContainer} {
            position: absolute;
            width: 100%;
            height: 100%;
            pointer-events: none;
            transform: rotate(-50deg);
            transform-origin: 0;

            > ${ShortStick} {
                position: absolute;
                left: 0;
                width: 2px;
                height: 0;
                background: ${theme.src.type[type$].form.radioColor};
                transition: height 60ms 40ms ease-in-out;
            }

            > ${LongStick} {
                position: absolute;
                left: 0;
                width: 0;
                height: 2px;
                background: ${theme.src.type[type$].form.radioColor};
                transition: width 40ms 0s ease-in-out;
            }
        }
    }

    &:checked + ${Rectangle} {
        border-color: ${theme.src.type[type$].main};
        background: ${theme.src.type[type$].main};

        > ${StickContainer} > {
            ${ShortStick} {
                transition: height 60ms 0s ease-in-out;
            }

            ${LongStick} {
                transition: width 40ms 60ms ease-in-out;
            }
        }
    }

    ${disabled ? `
        ~ * {
            opacity: ${theme.src.form.disabledOpacity};
            user-select: none;
        }
    ` : ""}
`)

export interface Props extends SizeProps {
    containerProps?: JSXElementProps
    inputProps?: JSXElementProps
    spanProps?: JSXElementProps
    children?: JSXChild
    type?: string
    fill?: boolean
    shape?: string
    value?: Value
    active?: Value[]
    setActive?: SetValue<Value[]>
    checked?: boolean
    setChecked?: SetValue<boolean>
}

const Checkbox:React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,containerProps=DV.JSX_ELEMENT_PROPS,inputProps=DV.JSX_ELEMENT_PROPS,spanProps=DV.JSX_ELEMENT_PROPS,children,type="primary",fill=true,shape="default",value,active,setActive,checked,setChecked}) => {
    const {checked$, change} = useMultiple({value,active,setActive,checked,setChecked,disabled:inputProps.disabled,onChange:inputProps.onChange})
    
    return (
        <RadioLabel {...containerProps} space={children ? true :false}>
            <Input {...inputProps} type="checkbox" size$={size} size$Sm={sizeSm} size$Md={sizeMd} size$Lg={sizeLg} size$Xl={sizeXl} type$={type} fill$={fill} shape={shape} checked={checked$} onChange={change} />
            <Rectangle>
                <StickContainer>
                    <ShortStick />
                    <LongStick />
                </StickContainer>
            </Rectangle>
            {children && <RadioSpan {...spanProps}>{children}</RadioSpan>}
        </RadioLabel>
    )
}

export default Checkbox