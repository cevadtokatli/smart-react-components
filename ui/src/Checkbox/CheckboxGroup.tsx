import extractElementProps from '@smart-react-components/core/element-props'
import intrinsicStyledProps, { IntrinsicStyledProps } from '@smart-react-components/core/element-props/intrinsic-styled-props'
import { PaletteProp, Partial, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import FormGroup from '../components/Form/FormGroup'
import { FormValue, OrderPosition } from '../types'

export { OrderPosition as Position } from '../types'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledProps {
  active?: FormValue[]
  children: JSX.Element[]
  isBlock?: boolean
  isOutline?: boolean
  isRequired?: boolean
  isSoft?: boolean
  name?: string
  palette?: PaletteProp
  position?: OrderPosition
  setActive?: SetState<FormValue[]>
  shape?: ShapeProp
  template?: JSX.Element
}

const CheckboxGroup: React.FC<Props> = props => (
  <FormGroup
    isBlock={props.isBlock}
    {...extractElementProps(props, [intrinsicStyledProps])}
  >
    { props.children.map((item, idx) => React.cloneElement(item, {
      key: item.key ?? idx,
      active: props.active,
      isBlock: props.isBlock,
      isOutline: props.isOutline,
      isRequired: props.isRequired,
      isSoft: props.isSoft,
      name: props.name,
      palette: props.palette,
      position: props.position,
      setActive: props.setActive,
      shape: props.shape,
      template: props.template,
      ...item.props,
    })) }
  </FormGroup>
)

export default CheckboxGroup
