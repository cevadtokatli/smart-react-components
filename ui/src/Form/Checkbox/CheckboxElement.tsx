import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { PaletteProp, ResponsiveProp, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import styled from 'styled-components'

const Container = styled.div(({ theme }) => `
  border: solid 2px ${theme.$.color.dynamic.accent};
  box-sizing: border-box;
  display: flex;
  height: 100%;
  justify-content: center;
  position: relative;
`)

const Rectangle = styled.div`
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
`

const Svg = styled.svg.attrs({
  children: <path d="M3.5,17.2l7.6,7.6L28.5,7.3" />,
  viewBox: '0 0 32 32',
})(({ theme }) => `
  flex: 0 0 auto;
  position: relative;

  > path {
    fill: transparent;
    stroke: ${theme.$.color.dynamic.background};
    stroke-width: 4px;
    stroke-dasharray: 36px;
    stroke-dashoffset: 36px;
    transition: stroke-dashoffset 200ms 0s ease-in-out;
  }
`)

interface Props extends
  StyledProps,
  ResponsiveProp<'checkboxSize', SizeProp> {
  isChecked: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
  shape: ShapeProp
}

export default styled(Div).attrs<Props>(({ children, isOutline, isSoft, palette, shape }) => ({
  children: (
    <>
      { children && children }
      <Container>
        <Rectangle />
        <Svg />
      </Container>
    </>
  ),
  getCheckboxSize: (v, t) => `
    height: ${t.$.size.form.checkbox[v]};
    width: ${t.$.size.form.checkbox[v]};
  `,
}))<Props>(({ theme, isChecked, isOutline, isSoft, palette, shape }: Props) => `
  position: relative;

  ${Container} {
    border-radius: ${theme.$.radius.form.checkbox[shape]};
  
    ${!isOutline
      ? `
        background: ${theme.$.color.dynamic.accent};
      `
      : ''
    }
  }

  ${Rectangle} {
    border-radius: ${theme.$.radius.form.checkbox[shape]};
  }

  ${isChecked
    ? `
      ${Rectangle} {
        background: ${!isSoft ? theme.$.palette[palette].main : theme.$.palette[palette].soft};
      }

      ${Svg} > path {
        stroke-dashoffset: 0;
      }
    `
    : ''
  }
`)
