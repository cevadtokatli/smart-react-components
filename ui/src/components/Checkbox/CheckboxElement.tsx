import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { PaletteProp, Partial, ResponsiveProp, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import styled from 'styled-components'

const Container = styled.div.attrs({
  className: 'src-checkbox-container',
})(({ theme }) => `
  border: solid 2px ${theme.$.color.dynamic.accent};
  box-sizing: border-box;
  display: flex;
  height: 100%;
  justify-content: center;
  position: relative;
`)

const Rectangle = styled.div.attrs({
  className: 'src-checkbox-rectangle',
})`
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
`

const Svg = styled.svg.attrs({
  children: <path d="M3.5,17.2l7.6,7.6L28.5,7.3" />,
  className: 'src-checkbox-svg',
  viewBox: '0 0 32 32',
})(({ theme }) => `
  flex: 0 0 auto;
  position: relative;

  > path {
    fill: transparent;
    stroke: ${theme.$.color.dynamic.background};
    stroke-width: 4px;
    stroke-dasharray: 36px;
    stroke-dashoffset: 36px;
    transition: stroke-dashoffset 200ms 0s ease-in-out;
  }
`)

const IndeterminateMark = styled.div.attrs({
  className: 'src-checkbox-indeterminate-mark',
})(({ theme }) => `
  background: ${theme.$.color.dynamic.background};
  height: 2px;
  margin-top: -1px;
  opacity: 0;
  position: absolute;
  trasnform: scaleX(0) rotate(45deg);
  transition-duration: 200ms;
  transition-property: opacity, transform;
  transition-timing-function: ease-in-out;
  top: 50%;
  width: 100%;
`)

export interface Props extends
  StyledProps,
  Partial<ResponsiveProp<'checkboxSize', SizeProp>> {
  isChecked?: boolean
  isIndeterminate?: boolean
  isOutline?: boolean
  isSoft?: boolean
  palette?: PaletteProp
  shape?: ShapeProp
}

export default styled(Div).attrs<Props>(({ children, isIndeterminate }) => ({
  children: (
    <>
      { children && children }
      <Container>
        <Rectangle />
        <Svg />
        { typeof isIndeterminate !== 'undefined' && <IndeterminateMark /> }
      </Container>
    </>
  ),
  getCheckboxSize: (v, t) => `
    height: ${t.$.size.checkbox[v]};
    width: ${t.$.size.checkbox[v]};
  `,
}))<Props>(({ theme, isChecked, isIndeterminate, isOutline, isSoft, palette, shape }: Props) => `
  position: relative;

  .src-checkbox-container {
    border-radius: ${theme.$.radius.checkbox[shape]};
    transition: border-color 200ms 0s ease-in-out;

    ${!isOutline
      ? `
        background: ${theme.$.color.dynamic.accent};
      `
      : ''
    }
  }

  .src-checkbox-rectangle {
    border-radius: ${theme.$.radius.checkbox[shape]};
  }

  ${isChecked
    ? `
      .src-checkbox-container {
        border-color: ${!isSoft ? theme.$.palette[palette].main : theme.$.palette[palette].soft};

        .src-checkbox-rectangle {
          background: ${!isSoft ? theme.$.palette[palette].main : theme.$.palette[palette].soft};
        }

        ${!isIndeterminate
          ? `
            .src-checkbox-svg > path {
              stroke-dashoffset: 0;
            }
          `
          : `
            .src-checkbox-indeterminate-mark {
              opacity: 1;
              transform: scaleX(1);
            }
          `
        }
      }
    `
    : ''
  }
`)
