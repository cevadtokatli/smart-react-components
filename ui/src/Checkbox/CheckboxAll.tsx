import extractElementProps from '@smart-react-components/core/element-props'
import intrinsicStyledCoreProps, { IntrinsicStyledCoreProps } from '@smart-react-components/core/element-props/intrinsic-styled-core-props'
import { ContentElement, PaletteProp, Partial, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import CheckboxElement from '../components/Checkbox/CheckboxElement'
import FormInlineLabel from '../components/Form/FormInlineLabel'
import HiddenInput from '../components/Form/HiddenInput'
import { FormValue, OrderPosition } from '../types'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledCoreProps {
  active?: FormValue[]
  items: FormValue[]
  isBlock?: boolean
  isDisabled?: boolean
  isOutline?: boolean
  isSoft?: boolean
  label?: ContentElement
  name?: string
  palette?: PaletteProp
  position?: OrderPosition
  setActive?: SetState<FormValue[]>
  shape?: ShapeProp
  template?: JSX.Element
}

const CheckboxAll = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => {
  const { isChecked, isIndeterminate } = React.useMemo(() => ({
    isChecked: props.active.length > 0,
    isIndeterminate: props.active.length > 0 && props.active.length < props.items.length,
  }), [props.active, props.items])

  const handleOnChange = () => {
    if (!props.isDisabled && props.setActive) {
      if (isChecked && !isIndeterminate) {
        props.setActive([])
      } else {
        props.setActive(props.items)
      }
    }
  }

  const input = (
    <HiddenInput
      checked={isChecked}
      onChange={handleOnChange}
      ref={forwardRef}
      type="checkbox"
      {...(props.isDisabled && { disabled: true })}
      {...(props.name && { name: props.name })}
    />
  )

  return (
    <FormInlineLabel
      formSize={props.size}
      formSizeSm={props.sizeSm}
      formSizeMd={props.sizeMd}
      formSizeLg={props.sizeLg}
      formSizeXl={props.sizeXl}
      isBlock={props.isBlock}
      isDisabled={props.isDisabled}
      labelPosition={props.position}
      {...extractElementProps(props, [intrinsicStyledCoreProps])}
    >
      { React.cloneElement(props.template, {
        checkboxSize: props.size,
        checkboxSizeSm: props.sizeSm,
        checkboxSizeMd: props.sizeMd,
        checkboxSizeLg: props.sizeLg,
        checkboxSizeXl: props.sizeXl,
        children: (
          <>
            {input}
            {props.template.props.children}
          </>
        ),
        isChecked,
        isIndeterminate,
        isOutline: props.isOutline,
        isSoft: props.isSoft,
        palette: props.palette,
        shape: props.shape,
      }) }
      { props.label && <span>{props.label}</span> }
    </FormInlineLabel>
  )
})

CheckboxAll.defaultProps = {
  palette: 'primary',
  position: OrderPosition.RIGHT,
  shape: 'rectangle',
  size: 'medium',
  template: <CheckboxElement />,
}

export default CheckboxAll
