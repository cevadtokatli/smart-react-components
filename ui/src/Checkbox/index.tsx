import extractElementProps from '@smart-react-components/core/element-props'
import intrinsicStyledProps, { IntrinsicStyledProps } from '@smart-react-components/core/element-props/intrinsic-styled-props'
import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { ContentElement, PaletteProp, Partial, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import FormInlineLabel from '../components/Form/FormInlineLabel'
import HiddenInput from '../components/Form/HiddenInput'
import { FormValue, OrderPosition } from '../types'
import CheckboxElement from './CheckboxElement'

export { OrderPosition as Position } from '../types'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledProps {
  active?: FormValue[]
  isBlock?: boolean
  isChecked?: boolean
  isDisabled?: boolean
  isOutline?: boolean
  isRequired?: boolean
  isSoft?: boolean
  label?: ContentElement
  name?: string
  palette?: PaletteProp
  position?: OrderPosition
  setActive?: SetState<FormValue[]>
  setChecked?: SetState<boolean>
  shape?: ShapeProp
  template?: JSX.Element
  value?: FormValue
}

const Checkbox = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => {
  const getChecked = () => {
    if (props.setChecked) {
      return props.isChecked
    }

    if (props.setActive) {
      return typeof props.active.find(v => v === props.value) !== 'undefined'
    }

    return false
  }

  const [isChecked, setChecked] = React.useState(() => getChecked())

  useChangeEffect(() => {
    setChecked(getChecked())
  }, [props.active, props.isChecked])

  const handleOnChange = () => {
    if (!props.isDisabled) {
      if (props.setChecked) {
        props.setChecked(!props.isChecked)
      } else if (props.setActive) {
        if (!props.active.includes(props.value)) {
          props.setActive([...props.active, props.value])
        } else {
          props.setActive(props.active.filter(item => item !== props.value))
        }
      }
    }
  }

  const input = (
    <HiddenInput
      checked={isChecked}
      onChange={handleOnChange}
      ref={forwardRef}
      type="checkbox"
      {...(props.isDisabled && { disabled: true })}
      {...(props.isRequired && { required: true })}
      {...(props.name && { name: props.name })}
    />
  )

  return (
    <FormInlineLabel
      formSize={props.size}
      formSizeSm={props.sizeSm}
      formSizeMd={props.sizeMd}
      formSizeLg={props.sizeLg}
      formSizeXl={props.sizeXl}
      isBlock={props.isBlock}
      isDisabled={props.isDisabled}
      labelPosition={props.position}
      {...extractElementProps(props, [intrinsicStyledProps])}
    >
      { React.cloneElement(props.template, {
        checkboxSize: props.size,
        checkboxSizeSm: props.sizeSm,
        checkboxSizeMd: props.sizeMd,
        checkboxSizeLg: props.sizeLg,
        checkboxSizeXl: props.sizeXl,
        children: (
          <>
            {input}
            {props.template.props.children}
          </>
        ),
        isChecked,
        isOutline: props.isOutline,
        isSoft: props.isSoft,
        palette: props.palette,
        shape: props.shape,
      }) }
      { props.label && <span>{props.label}</span> }
    </FormInlineLabel>
  )
})

Checkbox.defaultProps = {
  palette: 'primary',
  position: OrderPosition.RIGHT,
  shape: 'rectangle',
  size: 'medium',
  template: <CheckboxElement />,
}

export default Checkbox
