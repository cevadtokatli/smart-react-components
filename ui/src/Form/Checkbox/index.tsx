import extractElementProps from '@smart-react-components/core/element-props'
import intrinsicStyledProps, { IntrinsicStyledProps } from '@smart-react-components/core/element-props/intrinsic-styled-props'
import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { ContentElement, PaletteProp, Partial, ResponsiveProp, SetState, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import { FormValue, OrderPosition } from '../../types'
import FormHiddenInput from '../FormHiddenInput'
import FormInlineLabel from '../FormInlineLabel'
import CheckboxElement from './CheckboxElement'

export { OrderPosition as Position } from '../../types'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  IntrinsicStyledProps {
  active?: FormValue[]
  children?: ContentElement
  isBlock?: boolean
  isChecked?: boolean
  isDisabled?: boolean
  isOutline?: boolean
  isRequired?: boolean
  isSoft?: boolean
  name?: string
  palette?: PaletteProp
  position?: OrderPosition
  render?: (props: { isChecked: boolean }) => JSX.Element
  setActive?: SetState<FormValue[]>
  setChecked?: SetState<boolean>
  shape?: ShapeProp
  value?: FormValue
}

const Checkbox = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => {
  const getChecked = () => {
    if (props.setChecked) {
      return props.isChecked
    }

    if (props.setActive) {
      return typeof props.active.find(v => v === props.value) !== 'undefined'
    }

    return false
  }

  const [isChecked, setChecked] = React.useState(() => getChecked())

  useChangeEffect(() => {
    setChecked(getChecked())
  }, [props.active, props.isChecked])

  const handleOnChange = () => {
    if (!props.isDisabled) {
      if (props.setChecked) {
        props.setChecked(!props.isChecked)
      } else if (props.setActive) {
        if (!props.active.includes(props.value)) {
          props.setActive([...props.active, props.value])
        } else {
          props.setActive(props.active.filter(item => item !== props.value))
        }
      }
    }
  }

  const input = (
    <FormHiddenInput
      checked={isChecked}
      onChange={handleOnChange}
      ref={forwardRef}
      type="checkbox"
      {...(props.isDisabled && { disabled: true })}
      {...(props.isRequired && { required: true })}
      {...(props.name && { name: props.name })}
    />
  )

  return (
    <FormInlineLabel
      formSize={props.size}
      formSizeSm={props.sizeSm}
      formSizeMd={props.sizeMd}
      formSizeLg={props.sizeLg}
      formSizeXl={props.sizeXl}
      isBlock={props.isBlock}
      isDisabled={props.isDisabled}
      labelPosition={props.position}
      {...extractElementProps(props, [intrinsicStyledProps])}
    >
      { props.render
        ? (
          <>
            { props.render({ isChecked }) }
            { input }
          </>
          )
        : (
          <CheckboxElement
            checkboxSize={props.size}
            checkboxSizeSm={props.sizeSm}
            checkboxSizeMd={props.sizeMd}
            checkboxSizeLg={props.sizeLg}
            checkboxSizeXl={props.sizeXl}
            isChecked={isChecked}
            isOutline={props.isOutline}
            isSoft={props.isSoft}
            palette={props.palette}
            shape={props.shape}
          >
            { input }
          </CheckboxElement>
          )
      }
      { props.children && <span>{props.children}</span> }
    </FormInlineLabel>
  )
})

Checkbox.defaultProps = {
  palette: 'primary',
  position: OrderPosition.RIGHT,
  shape: 'rectangle',
  size: 'medium',
}

export default Checkbox
