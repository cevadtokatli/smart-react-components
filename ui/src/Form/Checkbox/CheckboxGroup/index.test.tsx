import { wrapTheme } from '@smart-react-components/core/test'
import { render } from '@testing-library/react'
import React from 'react'
import CheckboxGroup from './'
import Checkbox from '../'

describe('<CheckboxGroup />', () => {
  it('should render component', () => {
    const { asFragment } = render(wrapTheme(
      <CheckboxGroup active={['item-1']} setActive={jest.fn()}>
        <Checkbox value="item-1">Item 1</Checkbox>
        <Checkbox value="item-2">Item 2</Checkbox>
        <Checkbox value="item-3">Item 3</Checkbox>
      </CheckboxGroup>
    ))
    expect(asFragment()).toMatchSnapshot()
  })
})
