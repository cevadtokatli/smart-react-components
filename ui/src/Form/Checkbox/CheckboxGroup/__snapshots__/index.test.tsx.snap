// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxGroup /> should render component 1`] = `
<DocumentFragment>
  .c3 {
  height: 16px;
  width: 16px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c5 {
  height: 100%;
  left: 0;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  top: 0;
  width: 100%;
}

.c1 {
  margin: 1rem .5rem;
}

.c1 > :last-child {
  font-size: 1rem;
  margin-left: 0.46874rem;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: default;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
}

.c2 > :first-child {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c2 > :last-child {
  color: #323a46;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  font-family: inherit;
  word-break: break-word;
  word-wrap: break-word;
  -webkit-tap-highlight-color: transparent;
}

.c7 {
  border: solid 2px #dbdbdb;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.c9 {
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  -webkit-transition: background 200ms 0s ease-in-out;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
}

.c11 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  position: relative;
}

.c11 > path {
  fill: transparent;
  stroke: #fff;
  stroke-width: 4px;
  stroke-dasharray: 36px;
  stroke-dashoffset: 36px;
  -webkit-transition: stroke-dashoffset 200ms 0s ease-in-out;
  transition: stroke-dashoffset 200ms 0s ease-in-out;
}

.c4 {
  position: relative;
}

.c4 .c6 {
  border-radius: 2px;
  background: #dbdbdb;
}

.c4 .c8 {
  border-radius: 2px;
}

.c4 .c8 {
  background: #0665d0;
}

.c4 .c10 > path {
  stroke-dashoffset: 0;
}

.c12 {
  position: relative;
}

.c12 .c6 {
  border-radius: 2px;
  background: #dbdbdb;
}

.c12 .c8 {
  border-radius: 2px;
}

<div
    class=" c0"
  >
    <label
      class="c1 c2"
    >
      <div
        class="c3 c4"
        shape="rectangle"
      >
        <input
          checked=""
          class="c5"
          type="checkbox"
        />
        <div
          class="c6 c7"
        >
          <div
            class="c8 c9"
          />
          <svg
            class="c10 c11"
            viewBox="0 0 32 32"
          >
            <path
              d="M3.5,17.2l7.6,7.6L28.5,7.3"
            />
          </svg>
        </div>
      </div>
      <span>
        Item 1
      </span>
    </label>
    <label
      class="c1 c2"
    >
      <div
        class="c3 c12"
        shape="rectangle"
      >
        <input
          class="c5"
          type="checkbox"
        />
        <div
          class="c6 c7"
        >
          <div
            class="c8 c9"
          />
          <svg
            class="c10 c11"
            viewBox="0 0 32 32"
          >
            <path
              d="M3.5,17.2l7.6,7.6L28.5,7.3"
            />
          </svg>
        </div>
      </div>
      <span>
        Item 2
      </span>
    </label>
    <label
      class="c1 c2"
    >
      <div
        class="c3 c12"
        shape="rectangle"
      >
        <input
          class="c5"
          type="checkbox"
        />
        <div
          class="c6 c7"
        >
          <div
            class="c8 c9"
          />
          <svg
            class="c10 c11"
            viewBox="0 0 32 32"
          >
            <path
              d="M3.5,17.2l7.6,7.6L28.5,7.3"
            />
          </svg>
        </div>
      </div>
      <span>
        Item 3
      </span>
    </label>
  </div>
</DocumentFragment>
`;
